google/cloud/forseti/scanner/audit/kms_rules_engine.py

```diff
@@ -346,34 +338,107 @@ def __init__(self, rule_name, rule_index, rule):
         self.rule_index = rule_index
         self.rule = rule
 
-    def is_key_rotated(self, creation_time, scan_time):
-        """Check if the key has been rotated within the time speciifed in the
-         policy.
+    def find_match_rotation_period(self, key, rotation_period, mode):
+        """Check if there is a match for this rule rotation period against the
+        given resource.
 
         Args:
-            creation_time (datetime): The time at which the primary version of
-            the key was created.
-            scan_time (datetime): Snapshot timestamp.
+            key (Resource): The resource to check for a match.
+            mode (string): The mode specified in the rule.
+            rotation_period (string): The cut off rotation schedule of crypto
+            key specified in rule file.
 
         Returns:
-            bool: Returns true if key was rotated within the time specified.
+            bool: Returns true if a match is found.
         """
-        crypto_key = self.rule['key']
-        for key_data in crypto_key:
-            key_rotation_period = key_data.get('rotation_period')
         LOGGER.debug('Formatting rotation time...')
+        creation_time = key.primary_version.get('createTime')
+        scan_time = date_time.get_utc_now_datetime()
         last_rotation_time = creation_time[:-5]
         formatted_last_rotation_time = datetime.datetime.strptime(
             last_rotation_time, string_formats.TIMESTAMP_MICROS)
         days_since_rotated = (scan_time - formatted_last_rotation_time).days
+        print('days:', days_since_rotated)
+        if mode == BLACKLIST and days_since_rotated > rotation_period:
+            return True
+        elif mode == WHITELIST and days_since_rotated <= rotation_period:
+            return True
+        return False
+
+    @classmethod
+    def find_match_algorithms(cls, key, rule_algorithms):
+        """Check if there is a match for this rule algorithm against the given
+        resource.
+
+        Args:
+            key (Resource): The resource to check for a match.
+            rule_algorithms (string): The algorithms of this rule.
 
-        if days_since_rotated > key_rotation_period:
-            return False
+        Returns:
+            bool: Returns true if a match is found.
+        """
+        LOGGER.debug('Finding rule algorithm in key algorithm')
+        key_algorithm = key.primary_version.get('algorithm')
+        for algorithm in rule_algorithms:
+            if key_algorithm == algorithm:
+                return True
+        return False
+
+    def find_match_protection_level(self, key, rule_protection_level):
```

Comment:
Should rule_protection_level be a list, like algorithms?

Reply:
https://cloud.google.com/kms/docs/reference/rest/v1/ProtectionLevel

Technically there are only 2 valid values, either of which can be used at a time. Hence it is not a list.